Userland/Libraries/LibC/string.cpp

```diff
@@ -90,23 +90,24 @@ char* strndup(const char* str, size_t maxlen)
 // https://pubs.opengroup.org/onlinepubs/9699919799/functions/strcmp.html
 int strcmp(const char* s1, const char* s2)
 {
-    while (*s1 == *s2++)
-        if (*s1++ == 0)
-            return 0;
-    return *(const unsigned char*)s1 - *(const unsigned char*)--s2;
+    return strncmp(s1, s2, SIZE_MAX);
 }
 
 // https://pubs.opengroup.org/onlinepubs/9699919799/functions/strncmp.html
 int strncmp(const char* s1, const char* s2, size_t n)
 {
-    if (!n)
-        return 0;
-    do {
-        if (*s1 != *s2++)
-            return *(const unsigned char*)s1 - *(const unsigned char*)--s2;
-        if (*s1++ == 0)
+    const unsigned char* u1 = reinterpret_cast<const unsigned char*>(s1);
+    const unsigned char* u2 = reinterpret_cast<const unsigned char*>(s2);
+
+    for (size_t i = 0; i < n; i += 1) {
+        if (u1[i] < u2[i])
+            return -1;
+        if (u1[i] > u2[i])
+            return 1;
```

Comment:
I didn't do that because of underflow, is that not a problem here? e.g. if u1[i] == 1 and u2[i] == 250, I'd assume this would underflow and return 7 even though it should return a negative value because 1 < 250. Or are these automatically extended to int here? I never actually looked up how/when C/C++ auto extends integer types.

Reply:
cpp usually auto extents after the operation, but clangd will warn about that
to avoid this, cast one operand before the operation to the output type

Reply:
Oh wow, sometimes I don't think of the most obvious solutions. Thanks!